src/Uno.UI.Toolkit/Diagnostics/DiagnosticsOverlay.cs

```diff
@@ -0,0 +1,225 @@
+			{
+				lock (overlay._updateGate)
+				{
+					overlay._dispatcher = dispatcher;
+					overlay._updateCoordinator = dispatcher is null ? null : new Context(dispatcher);
+
+					// Clean all dispatcher bound state
+					overlay._overlayHost = null;
+					overlay._overlayPanel = null;
+					foreach (var element in overlay._elements.Values)
+					{
+						element.Dispose();
+					}
+					overlay._elements.Clear();
+				}
+			}
+			overlay.EnqueueUpdate();
+		};
+	}
+
+	public bool IsVisible
+	{
+		get => _isVisible;
+		set
+		{
+			_isVisible = value;
+			EnqueueUpdate(forceUpdate: !value); // For update when hiding.
+		}
+	}
+
+	/// <summary>
+	/// Add a UI diagnostic element to this overlay.
+	/// </summary>
+	/// <remarks>This will make this overlay <see cref="IsVisible"/> = true.</remarks>
+	public void Add(string name, UIElement preview, Func<UIElement>? details = null)
+		=> Add(new DiagnosticView(name, _ => preview, (_, ct) => new(details?.Invoke())));
+
+	/// <summary>
+	/// Add a UI diagnostic element to this overlay.
+	/// </summary>
+	/// <remarks>This will make this overlay <see cref="IsVisible"/> = true.</remarks>
+	/// <param name="provider">The provider to add.</param>
+	public void Add(IDiagnosticViewProvider provider)
+	{
+		lock (_updateGate)
+		{
+			_localProviders.Add(provider);
+		}
+
+		EnqueueUpdate(); // Making IsVisible = true wil (try) to re-enqueue the update, but safer to keep it here anyway.
+		IsVisible = true;
+	}
+
+	private void EnqueueUpdate(bool forceUpdate = false)
+	{
+		var dispatcher = _dispatcher;
+		if ((!_isVisible && !forceUpdate) || dispatcher is null || Interlocked.CompareExchange(ref _updateEnqueued, 1, 0) is not 0)
+		{
+			return;
+		}
+
+		dispatcher.TryEnqueue(() =>
+		{
+			_updateEnqueued = 0;
+
+			if (!_isVisible)
+			{
+				if (_overlayHost is { } host)
+				{
+					ShowHost(host, false);
+				}
+
+				return;
+			}
+
+			lock (_updateGate)
+			{
+				var providers = DiagnosticViewRegistry
+					.Registrations
+					.Where(ShouldMaterialize)
+					.Select(reg => reg.Provider)
+					.Concat(_localProviders)
+					.Distinct()
+					.ToList();
+
+				var panel = _overlayPanel ??= CreatePanel();
+				var host = _overlayHost ??= CreateHost(_root, panel);
+
+				foreach (var provider in providers)
+				{
+					if (!_elements.ContainsKey(provider))
+					{
+						var element = new DiagnosticElement(this, provider, _updateCoordinator!);
+						_elements[provider] = element;
+
+						panel.Children.Add(element.Preview);
+					}
+				}
+
+				ShowHost(host, true);
+			}
+		});
+	}
+
+	private static StackPanel CreatePanel()
+	{
+		var panel = new StackPanel
+		{
+			BorderThickness = new Thickness(1),
+			BorderBrush = new SolidColorBrush(Colors.Black),
+			Background = new SolidColorBrush(Colors.DarkGray),
+			Orientation = Orientation.Horizontal,
+			Padding = new Thickness(3),
+			Spacing = 3,
+			ManipulationMode = ManipulationModes.TranslateX | ManipulationModes.TranslateY
+		};
```

Comment:
Will it need to be styled?

Reply:
IMO no : the diagnostic overlay should be application's style independent ... and actually it has to prevent implicit styles of the application to impact it.